lib/galaxy/tools/verify/__init__.py

```diff
@@ -205,12 +206,21 @@ def get_lines_diff(diff):
             compressed_formats = None
         else:
             compressed_formats = []
-
-        # Open expected contents and history data in binary mode and run lines through unicodify,
-        # which will replace non utf-8 characters.
-        local_file = [unicodify(l) for l in get_fileobj(file1, mode='rb', compressed_formats=compressed_formats)]
-        history_data = [unicodify(l) for l in get_fileobj(file2, mode='rb', compressed_formats=compressed_formats)]
-        is_pdf = file1.endswith('.pdf') or file2.endswith('.pdf')
+        is_pdf = False
+        try:
+            with get_fileobj(file2, compressed_formats=compressed_formats) as fh:
+                history_data = fh.readlines()
+            with get_fileobj(file1, compressed_formats=compressed_formats) as fh:
+                local_file = fh.readlines()
+        except UnicodeDecodeError:
+            if file1.endswith('.pdf') or file2.endswith('.pdf'):
+                is_pdf = True
+                # Replace non-Unicode characters using unicodify(),
+                # difflib.unified_diff doesn't work on list of bytes
+                history_data = [unicodify(l) for l in get_fileobj(file2, mode='rb', compressed_formats=compressed_formats)]
+                local_file = [unicodify(l) for l in get_fileobj(file1, mode='rb', compressed_formats=compressed_formats)]
+            else:
+                raise AssertionError("Binary data detected, not displaying diff")
```

Comment:
This wouldn't work for the last example in the unit test if the extension isn't `.pdf`. This effectively makes non-utf8 encoded text uncomparable using `lines_diff`.  And it used to be comparable prior to our py3 porting.

Reply:
Actually that's OK, we already failed the lines diff at this point ...